src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1702,145 +1854,214 @@ final byte getCoder() {
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a string
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
+     */
+    private static byte[] appendChars(byte[] value, byte coder, int count, String s, int off, int end) {
+        if (isLatin1(coder)) {
             if (s.isLatin1()) {
-                System.arraycopy(s.value(), off, this.value, this.count, end - off);
+                System.arraycopy(s.value(), off, value, count, end - off);
             } else {
                 // We might need to inflate, but do it as late as possible since
                 // the range of characters we're copying might all be latin1
-                byte[] val = this.value;
                 for (int i = off, j = count; i < end; i++) {
                     char c = s.charAt(i);
                     if (StringLatin1.canEncode(c)) {
-                        val[j++] = (byte) c;
+                        value[j++] = (byte) c;
                     } else {
-                        count = j;
-                        inflate();
-                        System.arraycopy(s.value(), i << UTF16, this.value, j << UTF16, (end - i) << UTF16);
-                        count += end - i;
-                        return;
+                        value = inflateToUTF16(value, j);
+                        System.arraycopy(s.value(), i << UTF16, value, j << UTF16, (end - i) << UTF16);
+                        return value;
                     }
                 }
             }
         } else if (s.isLatin1()) {
-            StringUTF16.putCharsSB(this.value, this.count, s, off, end);
+            StringUTF16.putCharsSB(value, count, s, off, end);
         } else { // both UTF16
-            System.arraycopy(s.value(), off << UTF16, this.value, this.count << UTF16, (end - off) << UTF16);
+            System.arraycopy(s.value(), off << UTF16, value, count << UTF16, (end - off) << UTF16);
         }
-        count += end - off;
+        return value;
     }
 
-    private final void appendChars(CharSequence s, int off, int end) {
-        if (isLatin1()) {
-            byte[] val = this.value;
+    /**
+     * {@return buffer with new characters appended, possibly inflated}
+     * The buffer may need to be inflated if any character does not fit;
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a string
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
```

Comment:
It might be helpful to somehow clarify that, in the case of appending Latin1 characters to a Latin1 buffer, it is expected/required that `value` is already of sufficient size.